File: frontend/src/api/rbac.ts
import { http, unwrap } from './http'
import type { PageResult } from '../types/workover'

export interface UserItem {
  id: number
  username: string
  full_name: string
  department?: string | null
  mobile?: string | null
  email?: string | null
  is_active: boolean
  is_superuser: boolean
  role_ids: number[]
  created_at: string
}

export interface RoleItem {
  id: number
  name: string
  code: string
  description?: string | null
  is_active: boolean
  menu_ids: number[]
  permission_ids: number[]
}

export interface PermissionItem {
  id: number
  name: string
  code: string
  path: string
  method: string
  description?: string | null
  is_active: boolean
}

export interface MenuItem {
  id: number
  parent_id?: number | null
  title: string
  route_name: string
  route_path: string
  component?: string | null
  icon?: string | null
  sort_order: number
  is_visible: boolean
  is_active: boolean
  meta: Record<string, unknown>
  children?: MenuItem[]
}

export interface OperationLogItem {
  id: number
  user_id?: number | null
  username?: string | null
  ip_address?: string | null
  method: string
  path: string
  operation?: string | null
  status_code?: number | null
  response_message?: string | null
  created_at: string
}

export function listUsers() {
  return unwrap<UserItem[]>(http.get('/users'))
}

export function createUser(payload: {
  username: string
  password: string
  full_name: string
  department?: string
  mobile?: string
  email?: string
  is_active: boolean
  role_ids: number[]
}) {
  return unwrap<UserItem>(http.post('/users', payload))
}

export function setUserActive(id: number, isActive: boolean) {
  return unwrap<UserItem>(http.patch(`/users/${id}/active`, null, { params: { is_active: isActive } }))
}

export function deleteUser(id: number) {
  return unwrap<null>(http.delete(`/users/${id}`))
}

export function assignUserRoles(id: number, roleIds: number[]) {
  return unwrap<UserItem>(http.patch(`/users/${id}/roles`, { ids: roleIds }))
}

export function listRoles() {
  return unwrap<RoleItem[]>(http.get('/roles'))
}

export function listPermissions() {
  return unwrap<PermissionItem[]>(http.get('/permissions'))
}

export function listMenus() {
  return unwrap<MenuItem[]>(http.get('/menus'))
}

export function assignRolePermissions(id: number, permissionIds: number[]) {
  return unwrap<RoleItem>(http.patch(`/roles/${id}/permissions`, { ids: permissionIds }))
}

export function listOperationLogs(query: { page?: number; page_size?: number } = {}) {
  return unwrap<PageResult<OperationLogItem>>(http.get('/operation-logs', { params: query }))
}

File: app/schemas/rbac.py
from datetime import datetime
from typing import Any

from pydantic import BaseModel, ConfigDict, Field, field_validator


class IdsPayload(BaseModel):
    ids: list[int] = Field(default_factory=list)


class UserCreate(BaseModel):
    username: str = Field(min_length=1, max_length=64)
    password: str = Field(min_length=8, max_length=128)
    full_name: str = Field(min_length=1, max_length=64)
    department: str | None = Field(default=None, max_length=128)
    mobile: str | None = Field(default=None, max_length=32)
    email: str | None = Field(default=None, max_length=128)
    is_active: bool = True
    role_ids: list[int] = Field(default_factory=list)
    extra_config: dict[str, Any] = Field(default_factory=dict)

    @field_validator("password")
    @classmethod
    def validate_password_complexity(cls, value: str) -> str:
        checks = (
            any(char.islower() for char in value),
            any(char.isupper() for char in value),
            any(char.isdigit() for char in value),
            any(not char.isalnum() for char in value),
        )
        if not all(checks):
            raise ValueError("Password must include uppercase, lowercase, number, and special character")
        return value


class UserUpdate(BaseModel):
    full_name: str = Field(min_length=1, max_length=64)
    department: str | None = Field(default=None, max_length=128)
    mobile: str | None = Field(default=None, max_length=32)
    email: str | None = Field(default=None, max_length=128)
    is_active: bool = True
    extra_config: dict[str, Any] = Field(default_factory=dict)


class UserOut(BaseModel):
    id: int
    username: str
    full_name: str
    department: str | None = None
    mobile: str | None = None
    email: str | None = None
    is_active: bool
    is_superuser: bool
    role_ids: list[int] = []
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


class RoleCreate(BaseModel):
    name: str = Field(min_length=1, max_length=64)
    code: str = Field(min_length=1, max_length=64)
    description: str | None = Field(default=None, max_length=255)
    is_active: bool = True
    extra_config: dict[str, Any] = Field(default_factory=dict)


class RoleUpdate(RoleCreate):
    pass


class RoleOut(BaseModel):
    id: int
    name: str
    code: str
    description: str | None = None
    is_active: bool
    menu_ids: list[int] = []
    permission_ids: list[int] = []

    model_config = ConfigDict(from_attributes=True)


class MenuCreate(BaseModel):
    parent_id: int | None = None
    title: str = Field(min_length=1, max_length=64)
    route_name: str = Field(min_length=1, max_length=128)
    route_path: str = Field(min_length=1, max_length=255)
    component: str | None = Field(default=None, max_length=255)
    icon: str | None = Field(default=None, max_length=64)
    sort_order: int = 0
    is_visible: bool = True
    is_active: bool = True
    meta: dict[str, Any] = Field(default_factory=dict)


class MenuUpdate(MenuCreate):
    pass


class MenuOut(BaseModel):
    id: int
    parent_id: int | None = None
    title: str
    route_name: str
    route_path: str
    component: str | None = None
    icon: str | None = None
    sort_order: int
    is_visible: bool
    is_active: bool
    meta: dict[str, Any]
    children: list["MenuOut"] = []

    model_config = ConfigDict(from_attributes=True)


class PermissionCreate(BaseModel):
    name: str = Field(min_length=1, max_length=64)
    code: str = Field(min_length=1, max_length=128)
    path: str = Field(min_length=1, max_length=255)
    method: str = Field(min_length=1, max_length=16)
    description: str | None = Field(default=None, max_length=255)
    is_active: bool = True
    extra_config: dict[str, Any] = Field(default_factory=dict)


class PermissionUpdate(PermissionCreate):
    pass


class PermissionOut(BaseModel):
    id: int
    name: str
    code: str
    path: str
    method: str
    description: str | None = None
    is_active: bool

    model_config = ConfigDict(from_attributes=True)


class OperationLogOut(BaseModel):
    id: int
    trace_id: str | None = None
    user_id: int | None = None
    username: str | None = None
    ip_address: str | None = None
    method: str
    path: str
    operation: str | None = None
    status_code: int | None = None
    response_message: str | None = None
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


class SupportMetricOut(BaseModel):
    name: str
    status: str
    value: str | int | float | None = None
    description: str | None = None


class SystemSupportOverviewOut(BaseModel):
    runtime_monitoring: list[SupportMetricOut] = []
    security_controls: list[SupportMetricOut] = []
    audit_traceability: list[SupportMetricOut] = []
    backup_recovery: list[SupportMetricOut] = []
    message_alerts: list[SupportMetricOut] = []
    data_scope: list[SupportMetricOut] = []


class ApprovalLogOut(BaseModel):
    id: int
    business_type: str
    business_id: int
    node_code: str
    action: str
    comment: str | None = None
    operator_id: int | None = None
    operator_ip: str | None = None
    before_snapshot: dict[str, Any] | None = None
    after_snapshot: dict[str, Any] | None = None
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)
